cc/gtp_player.cc

```diff
@@ -0,0 +1,323 @@
+  if (cmd == "clear_board") {
+    return HandleClearBoard(cmd, args);
+  } else if (cmd == "echo") {
+    return HandleEcho(cmd, args);
+  } else if (cmd == "final_score") {
+    return HandleFinalScore(cmd, args);
+  } else if (cmd == "gamestate" || cmd == "mg_gamestate") {
+    return HandleGamestate(cmd, args);
+  } else if (cmd == "genmove" || cmd == "mg_genmove") {
+    return HandleGenmove(cmd, args);
+  } else if (cmd == "play") {
+    return HandlePlay(cmd, args);
+  } else if (cmd == "readouts") {
+    return HandleReadouts(cmd, args);
+  } else if (cmd == "report_search_interval") {
+    return HandleReportSearchInterval(cmd, args);
+  }
+  std::cerr << "ERROR: unrecognized GTP command \"" << cmd << "\"\n";
+  return "";
+}
+
+std::string GtpPlayer::HandleClearBoard(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  if (!CheckArgsExact(cmd, 0, args)) {
+    return "";
+  }
+
+  NewGame();
+
+  return "";
+}
+
+std::string GtpPlayer::HandleEcho(absl::string_view cmd,
+                                  const std::vector<absl::string_view>& args) {
+  return absl::StrJoin(args, " ");
+}
+
+std::string GtpPlayer::HandleFinalScore(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  if (!CheckArgsExact(cmd, 0, args)) {
+    return "";
+  }
+  return absl::StrJoin(args, " ");
+}
+
+std::string GtpPlayer::HandleGamestate(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  if (!CheckArgsExact(cmd, 0, args)) {
+    return "";
+  }
+
+  const auto& position = root()->position;
+
+  // board field.
+  std::ostringstream oss;
+  for (const auto& stone : position.stones()) {
+    char ch;
+    if (stone.color() == Color::kBlack) {
+      ch = 'X';
+    } else if (stone.color() == Color::kWhite) {
+      ch = 'O';
+    } else {
+      ch = '.';
+    }
+    oss << ch;
+  }
+  std::string board = oss.str();
+
+  // toPlay field.
+  std::string to_play = position.to_play() == Color::kBlack ? "Black" : "White";
+
+  // lastMove field.
+  std::string last_move;
+  if (!history().empty()) {
+    last_move = absl::StrCat("\"", history().back().c.ToKgs(), "\"");
+  } else {
+    last_move = "null";
+  }
+
+  // n field.
+  auto n = position.n();
+
+  // q field.
+  auto q = root()->parent != nullptr ? root()->parent->Q() : 0;
+
+  // clang-format off
+    std::cerr << "mg-gamestate: {"
+              << "\"board\":\"" << board << "\", "
+              << "\"toPlay\":\"" << to_play << "\", "
+              << "\"lastMove\":" << last_move << ", "
+              << "\"n\":" << n << ", "
+              << "\"q\":" << q
+              << "}" << std::endl;
+  // clang-format on
+  return "";
+}
+
+std::string GtpPlayer::HandleGenmove(
+    absl::string_view cmd, const std::vector<absl::string_view>& args) {
+  if (!CheckArgsRange(cmd, 0, 1, args)) {
+    return "";
+  }
+
+  auto c = SuggestMove(num_readouts_);
+  std::cerr << root()->Describe() << std::endl;
+  if (ShouldResign()) {
```

Comment:
out of ignorance does pass also need special handling

Reply:
no. resign doesn't need special handling now either :)